[Python, Visually](https://johnfisher-ai.github.io/Python-Visual-Guides/) &nbsp;&rsaquo;&nbsp; [Python from the Start](https://johnfisher-ai.github.io/Python-Visual-Guides/python-from-the-start.html)

# Debugging &middot; Solutions


One way to do each task. Not the only way. If yours runs and does what was asked, yours is
right too.


**1.** Create `count = 12` and `label = "items"`, then print both using the `=` form.


In [1]:
count = 12
label = "items"

print(f"{count=}")
print(f"{label=}")


count=12
label='items'


Note that `label` prints with quotes. The `=` form uses `repr`, which is what you want when
debugging: it distinguishes the text `12` from the number `12`.


**2.** Given `first = "yes"` and `second = "yes "`, print both with `repr` and print
whether they are equal.


In [2]:
first = "yes"
second = "yes "

print(f"{first!r}")
print(f"{second!r}")
print(f"{(first == second)=}")

# second has a trailing space, which print would not have shown.


'yes'
'yes '
(first == second)=False


This is the most common form of "but they look the same". Data read from a file or a form
arrives with whitespace attached far more often than not, which is why `.strip()` exists.


**3.** Given `mixed = [1, "2", 3.0, None]`, print each item with its type name.


In [3]:
mixed = [1, "2", 3.0, None]

for item in mixed:
    print(f"{item!r:<8} {type(item).__name__}")


1        int
'2'      str
3.0      float
None     NoneType


`!r` keeps `None` and `"2"` distinguishable from `1` and `3.0` in the output, and the `:<8`
lines the types up so the column can be scanned.


**4.** Given `values = ["5", "6", "seven", "8"]`, print the index and value of every item
that cannot be converted with `int`.


In [4]:
values = ["5", "6", "seven", "8"]

for i, v in enumerate(values):
    try:
        int(v)
    except ValueError as e:
        print(f"index {i}: {v!r} -> {e}")


index 2: 'seven' -> invalid literal for int() with base 10: 'seven'


The loop reports the position as well as the value, which matters when the same bad value
appears twice and only one of them is a problem.


**5.** Write `first_item(items)` with an `assert` that the list is not empty.


In [5]:
def first_item(items):
    assert len(items) > 0, "first_item was given an empty list"
    return items[0]

print(first_item(["a", "b"]))


a


In [6]:
print(first_item([]))


AssertionError: first_item was given an empty list

Without the assert this would have been `IndexError: list index out of range`, which is accurate
and says nothing about which function was misused or why.

The message is the part that earns its place. `assert len(items) > 0` on its own would fail just
as correctly and tell you nothing.


**6.** Compute the average of the items that **are** numbers, and print how many were
skipped.


In [7]:
numbers = []
skipped = 0

for v in values:
    try:
        numbers.append(int(v))
    except ValueError:
        skipped += 1

print(f"{numbers=}")
print(f"{skipped=}")
print("average:", sum(numbers) / len(numbers))


numbers=[5, 6, 8]
skipped=1
average: 6.333333333333333


The denominator is `len(numbers)`, not `len(values)`. Dividing by four here would give `4.75`
instead of `6.33`, which is the exact mistake the last section of the notebook demonstrates.

Reporting `skipped` is the other half. An average built from three of four values is a different
claim from one built on all four, and the output should say which it is.


---

&#8592; **Back to:** [Debugging](https://colab.research.google.com/github/johnfisher-ai/Python-Visual-Guides/blob/main/notebooks/python-from-the-start/20-debugging.ipynb)  &nbsp;&middot;&nbsp;  [Python from the Start Notebooks](https://johnfisher-ai.github.io/Python-Visual-Guides/python-from-the-start.html)
